source/extensions/filters/network/postgres_proxy/postgres_message.h

```diff
@@ -0,0 +1,270 @@
+public:
+  /**
+   * See above for parameter and return value description.
+   */
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left);
+  std::string toString() const;
+
+private:
+  std::vector<uint8_t> value_;
+};
+
+// VarByteN represents the structure consisting of 4 bytes of length
+// indicating how many bytes follow.
+// In Postgres documentation it is described as:
+// - Int32
+//   The number of bytes in the structure (this count does not include itself). Can be
+//   zero. As a special case, -1 indicates a NULL (no result). No value bytes follow in the NULL
+// case.
+//
+// - ByteN
+// The sequence of bytes representing the value. Bytes are present only when length has a positive
+// value.
+class VarByteN {
+public:
+  /**
+   * See above for parameter and return value description.
+   */
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left);
+  std::string toString() const;
+
+private:
+  int32_t len_;
+  std::vector<uint8_t> value_;
+};
+
+// Array contains one or more values of the same type.
+template <typename T> class Array {
+public:
+  /**
+   * See above for parameter and return value description.
+   */
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    // First read the 16 bits value which indicates how many
+    // elements there are in the array.
+    if (((data.length() - pos) < sizeof(uint16_t)) || (left < sizeof(uint16_t))) {
+      return false;
+    }
+    const uint16_t num = data.peekBEInt<uint16_t>(pos);
+    pos += sizeof(uint16_t);
+    left -= sizeof(uint16_t);
+    if (num != 0) {
+      for (auto i = 0; i < num; i++) {
+        auto item = std::make_unique<T>();
+        if (!item->read(data, pos, left)) {
+          return false;
+        }
+        value_.push_back(std::move(item));
+      }
+    }
+    return true;
+  }
+
+  std::string toString() const {
+    std::string out = fmt::format("[Array of {}:{{", value_.size());
+
+    for (const auto& i : value_) {
+      absl::StrAppend(&out, i->toString());
+    }
+    absl::StrAppend(&out, "}]");
+
+    return out;
+  }
+
+private:
+  std::vector<std::unique_ptr<T>> value_;
+};
+
+// Repeated is a composite type used at the end of the message.
+// It indicates to read the value of the same type until the end
+// of the Postgres message.
+template <typename T> class Repeated {
+public:
+  /**
+   * See above for parameter and return value description.
+   */
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    if ((data.length() - pos) < left) {
+      return false;
+    }
+    // Read until nothing is left.
+    while (left != 0) {
+      auto item = std::make_unique<T>();
+      if (!item->read(data, pos, left)) {
+        return false;
+      }
+      value_.push_back(std::move(item));
+    }
+    return true;
+  }
+
+  std::string toString() const {
+    std::string out;
+
+    for (const auto& i : value_) {
+      absl::StrAppend(&out, i->toString());
```

Comment:
ditto -- comment why we don't need a delimiter between the stringified instances.